rust/kernel/thread.rs

```diff
@@ -0,0 +1,390 @@
+            name.as_ptr(),
+        ))?;
+
+        // Increases the refcount of the task, so that it won't go away if it
+        // `do_exit`.
+        //
+        // SAFETY: `task` is a proper pointer pointing to a newly created
+        // thread.
+        rust_helper_get_task_struct(task);
+
+        Ok(Thread { task })
+    }
+
+    /// Creates a new thread.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use kernel::thread::{Thread, ThreadFunctor};
+    /// use alloc::boxed::Box;
+    /// use core::fmt::Debug;
+    ///
+    /// struct SimpleBoxFunc;
+    ///
+    /// impl<T> ThreadFunc<Box<T>> for SimpleBoxFunc
+    /// where T: Send + Debug
+    /// {
+    ///     // `Box<T>` implements [`ThreadArg`]
+    ///     fn func(b: Box<T>) -> KernelResult {
+    ///         println!("I got {:?}", b);
+    ///     }
+    /// }
+    ///
+    /// let t = Thread::try_new::<_, SimpleBoxFunc>(
+    ///   cstr!("rust-thread"),
+    ///   Box::try_new(42)?
+    /// )?;
+    ///
+    /// t.wake_up();
+    /// ```
+    ///
+    /// # Context
+    ///
+    /// This function might sleep in `kthread_create_on_node` due to the memory
+    /// allocation and waiting for the completion, therefore do not call this
+    /// in atomic contexts (i.e. preemption-off contexts).
+    pub fn try_new<A, F: ThreadFunctor<A>>(name: CStr, arg: A) -> KernelResult<Self>
+    where
+        A: ThreadArg,
+    {
+        let data = arg.into_pointer();
+
+        // SAFETY: `bridge::<A, F>` is a proper function pointer to a C
+        // function, and [`ThreadArg::from_pointer`] will be used in it to
+        // consume the raw pointer in the new thread.
+        let result = unsafe { Self::try_new_extern(name, bridge::<A, F>, data as _) };
+
+        if let Err(e) = result {
+            // Creation fails, we need to consume the raw pointer `data` because
+            // there is no new thread to own the underlying object, we should
+            // let the current thread own it.
+            //
+            // SAFETY: We `from_pointer` back what we just `into_pointer`, and
+            // since the new thread is not created, so no one touches `data`.
+            unsafe {
+                A::from_pointer(data);
+            }
+
+            return Err(e);
+        }
+
+        result
+    }
+
+    /// Creates a new thread with a closure.
+    ///
+    /// # Memory Cost
+    ///
+    /// The implementation of this function currently need to allocate extra
+    /// dynamic memory (using [`Box`]), and do double boxing because
+    /// `Box<dyn FnOnce()>`. These "drawbacks" similarly exists in the
+    /// implementation of [`std::thread::spawn`]. Only use this when these
+    /// are acceptable, otherwise try [`Thread::try_new`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use kernel::thread::Thread;
+    /// use alloc::boxed::Box;
+    ///
+    /// let mut a = 1;
+    ///
+    /// let t = Thread::try_new_closure(
+    ///   cstr!("rust-thread"),
```

Comment:
Indentation? Please check with `rustfmt` if the two-spaces are correct.

Reply:
`make <...> rustfmtcheck` doesn't report anything. Actually I now make sure to run `rustfmtcheck` every time I push the code.  That said 4 spaces indentation does make more sense, I will fix this.

Reply:
`rustfmt` does not check inside comments AFAIK. In other words, what I meant is that you can try putting the code in a file/function, format it, and then copy the result into the comment :-)

Reply:
Actually, just checked -- it does support it, but it is not enabled by default. I will send a PR for it.

Reply:
Done: #163.

Reply:
Thanks, I will rebase, check and update.